### ISSUE TITLE
Machiney crash when non existing bin in filesystem

### ISSUE THREAD

**Issue by tigerfoot** [SUSE/machinery#2037]
Hello, I've found another way to crash machinery

```
Inspecting dumbo.disney.interne for os, packages, patterns, repositories, users, groups, services, config-files, changed-managed-files, unmanaged-files...                                                    
                                                                                                                                                                                                              
Note: There are filters being applied during inspection. (Use `--verbose` option to show the filters)                                                                                                         
                                                                                                                                                                                                              
                                                                                                                                                                                                              
Inspecting os...
 -> Found operating system 'openSUSE 13.1 (Bottle)' version '13.1 (Bottle)'.
Inspecting packages...
 -> Found 2331 packages.
Inspecting patterns...
 -> Found 11 patterns.
Inspecting repositories...
 -> Found 24 repositories.
Inspecting users...
 -> Found 47 users.
Inspecting groups...
 -> Found 68 groups.
Inspecting services...
 -> Found 188 services.
Inspecting config-files...
 -> Found 93 config files...Machinery experienced an unexpected error. Please file a bug report at: https://github.com/SUSE/machinery/issues/new
Execution of "ssh root@dumbo.disney.interne -o LogLevel\=ERROR LANGUAGE\= LC_ALL\=en_US.utf8 rpm -qf /usr/bin/uncompress" failed with status 1: error: file /usr/bin/uncompress: No such file or directory.

Error output:
error: file /usr/bin/uncompress: No such file or directory

Backtrace:
/usr/lib64/ruby/gems/2.2.0/gems/cheetah-0.5.0/lib/cheetah.rb:641:in `check_errors'
/usr/lib64/ruby/gems/2.2.0/gems/cheetah-0.5.0/lib/cheetah.rb:404:in `run'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/logged_cheetah.rb:23:in `run'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/remote_system.rb:92:in `run_command'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/rpm_database.rb:24:in `package_for_file_path'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/managed_files_database.rb:56:in `block in changed_files'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/managed_files_database.rb:49:in `each'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/managed_files_database.rb:49:in `inject'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/managed_files_database.rb:49:in `changed_files'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/plugins/config_files/config_files_inspector.rb:44:in `inspect'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/inspect_task.rb:87:in `block in build_description'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/inspect_task.rb:79:in `each'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/inspect_task.rb:79:in `build_description'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/inspect_task.rb:21:in `inspect_system'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/cli.rb:610:in `block (2 levels) in <class:Cli>'
/usr/lib64/ruby/gems/2.2.0/gems/gli-2.13.1/lib/gli/command_support.rb:126:in `call'
/usr/lib64/ruby/gems/2.2.0/gems/gli-2.13.1/lib/gli/command_support.rb:126:in `execute'
/usr/lib64/ruby/gems/2.2.0/gems/gli-2.13.1/lib/gli/app_support.rb:296:in `block in call_command'
/usr/lib64/ruby/gems/2.2.0/gems/gli-2.13.1/lib/gli/app_support.rb:309:in `call'
/usr/lib64/ruby/gems/2.2.0/gems/gli-2.13.1/lib/gli/app_support.rb:309:in `call_command'
/usr/lib64/ruby/gems/2.2.0/gems/gli-2.13.1/lib/gli/app_support.rb:83:in `run'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/bin/machinery:33:in `<top (required)>'
/usr/bin/machinery:23:in `load'
/usr/bin/machinery:23:in `<main>'
```

```
 rpm --verify gzip
missing     /usr/bin/uncompress (replaced)
```

So to reproduce and create a test, install gzip delete /usr/bin/uncompress 
run machinery remote -> crash

Machinery version : 1.18.0-1.1 under tumbleweed
Inspected computer : openSUSE 13.1 x86_64

**Comment by thardeck** [comment]
Thanks for reporting the issue. I could not reproduce it when `uncompress` is removed before inspection but when it is removed during inspection, shortly after finding all changed rpm files it will crash.
I will try to make this code path more robust.

**Comment by thardeck** [comment]
I have now tried to reproduced it by deleting a file marked as replaced (used by two conflicting packages) and it does not crash but marks it correctly as (replaced, deleted).

We had an issue where this case is not covered, but this was fixed before the release of 1.1.18.

So I am wondering if the file was deleted before the machinery inspection or during?
```
# Changed Managed Files [10.160.67.129] (2016-03-24 08:22:12)

  Files extracted: yes
  * /usr/bin/gem (replaced)
```

```
# Changed managed files [10.160.67.129] (2016-03-24 08:31:40)

  Files extracted: yes
  * /usr/bin/gem (replaced, deleted)
```

**Comment by tigerfoot** [comment]
Ok I've tried to reproduce here (directly on my TW workstation)
So I removed /usr/bin/uncompress
```
rm -f /usr/bin/uncompress
rpm -qf /usr/bin/uncompress
gzip-1.6-10.21.x86_64

rpm -qV gzip
missing     /usr/bin/uncompress
```
Afterward I run machinery inspect and got anoter backtrace for another missing file in the FS

```
Inspecting qt-kt for os, packages, patterns, repositories, users, groups, services, config-files, changed-managed-files, unmanaged-files...

Note: There are filters being applied during inspection. (Use `--verbose` option to show the filters)


Inspecting os...
 -> Found operating system 'openSUSE Tumbleweed' version '20160321'.
Inspecting packages...
 -> Found 3353 packages.
Inspecting patterns...
 -> Found 20 patterns.
Inspecting repositories...
 -> Found 26 repositories.
Inspecting users...
 -> Found 58 users.
Inspecting groups...
 -> Found 81 groups.
Inspecting services...
 -> Found 242 services.
Inspecting config-files...
 -> Found 74 config files...Machinery experienced an unexpected error. Please file a bug report at: https://github.com/SUSE/machinery/issues/new
Execution of "ssh root@qt-kt -o LogLevel\=ERROR LANGUAGE\= LC_ALL\=en_US.utf8 rpm -qf /usr/lib64/dri/vdpau_drv_video.so" failed with status 1: error: file /usr/lib64/dri/vdpau_drv_video.so: No such file or directory.

Error output:
error: file /usr/lib64/dri/vdpau_drv_video.so: No such file or directory

Backtrace:
/usr/lib64/ruby/gems/2.2.0/gems/cheetah-0.5.0/lib/cheetah.rb:641:in `check_errors'
/usr/lib64/ruby/gems/2.2.0/gems/cheetah-0.5.0/lib/cheetah.rb:404:in `run'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/logged_cheetah.rb:23:in `run'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/remote_system.rb:92:in `run_command'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/rpm_database.rb:24:in `package_for_file_path'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/managed_files_database.rb:56:in `block in changed_files'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/managed_files_database.rb:49:in `each'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/managed_files_database.rb:49:in `inject'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/managed_files_database.rb:49:in `changed_files'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/plugins/config_files/config_files_inspector.rb:44:in `inspect'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/inspect_task.rb:87:in `block in build_description'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/inspect_task.rb:79:in `each'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/inspect_task.rb:79:in `build_description'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/inspect_task.rb:21:in `inspect_system'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/lib/cli.rb:610:in `block (2 levels) in <class:Cli>'
/usr/lib64/ruby/gems/2.2.0/gems/gli-2.13.1/lib/gli/command_support.rb:126:in `call'
/usr/lib64/ruby/gems/2.2.0/gems/gli-2.13.1/lib/gli/command_support.rb:126:in `execute'
/usr/lib64/ruby/gems/2.2.0/gems/gli-2.13.1/lib/gli/app_support.rb:296:in `block in call_command'
/usr/lib64/ruby/gems/2.2.0/gems/gli-2.13.1/lib/gli/app_support.rb:309:in `call'
/usr/lib64/ruby/gems/2.2.0/gems/gli-2.13.1/lib/gli/app_support.rb:309:in `call_command'
/usr/lib64/ruby/gems/2.2.0/gems/gli-2.13.1/lib/gli/app_support.rb:83:in `run'
/usr/lib64/ruby/gems/2.2.0/gems/machinery-tool-1.18.0/bin/machinery:33:in `<top (required)>'
/usr/bin/machinery:23:in `load'
/usr/bin/machinery:23:in `<main>'
```

```
 zypper se -sf  /usr/lib64/dri/vdpau_drv_video.so
S | Name               | Type    | Version   | Arch   | Repository
--+--------------------+---------+-----------+--------+-----------
i | libva-vdpau-driver | package | 0.7.4-1.3 | x86_64 | oss
v | libva-vdpau-driver | package | 0.7.4-1.3 | i586   | oss
```
I guess one of my nvidia binary try kill that file :-(

Even if /usr/bin/uncompress is not present (missing) the next run was successfull
```
Inspecting os...
 -> Found operating system 'openSUSE Tumbleweed' version '20160321'.
Inspecting packages...
 -> Found 3353 packages.
Inspecting patterns...
 -> Found 20 patterns.
Inspecting repositories...
 -> Found 26 repositories.
Inspecting users...
 -> Found 58 users.
Inspecting groups...
 -> Found 81 groups.
Inspecting services...
 -> Found 242 services.
Inspecting config-files...
 -> Found 72 changed configuration files.
Inspecting changed-managed-files...
 -> Found 95 changed files.
Inspecting unmanaged-files...
 -> Found 3463 unmanaged files and trees.
```
show 
 * /usr/bin/uncompress (deleted)

Question : as rpm -qV state missing shouldn't machinery used the same words in place of deleted ?

**Comment by thardeck** [comment]
Thx for your quick response.

Is it possible that you have auto-update enabled and that the Nvidia driver was updated in the background during inspection?

I am not sure why we choose "deleted", I guess it depends on the perspective, from rpm or the user side.

**Comment by tigerfoot** [comment]
For the nvidia driver, they were installed last week during a few hours (rpm based on what we have for openSUSE) then removed as they were worse than nouveau.
So no modification occurs during the inspect timeframe

**Comment by cornelius** [comment]
I suppose this will be fixed by #2188
